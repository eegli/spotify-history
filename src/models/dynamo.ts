interface DynamoBaseDoc {
  TableName: string;
}

// Primary key for this table
// Optional sort key may be added here
interface DynamoBaseKeys {
  dateId: string;
}

// For GET operations
export interface DynamoRef extends DynamoBaseDoc {
  Key: DynamoBaseKeys;
}

// For PUT operations
export interface DynamoItem extends DynamoBaseDoc {
  Item: DynamoBaseKeys & DynamoRecordAttrs;
}

// All non-required item attributes
interface DynamoRecordAttrs {
  lastScrobbed: string;
  lastScrobbedString: string;
  songs?: any;
}
