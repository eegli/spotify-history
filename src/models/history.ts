import {
  attribute,
  hashKey,
  rangeKey,
  table,
} from '@aws/dynamodb-data-mapper-annotations';
import { CustomType } from '@aws/dynamodb-data-marshaller';
import { AttributeValue } from 'aws-sdk/clients/dynamodb';
import moment from 'moment';
import config from '../config';

// Define the properties of a song that is eventually saved to Dynamo
export interface HistoryElement {
  name: string;
  id: string;
  playedAt: string;
  artists: {
    artistName: string;
    artistId: string;
    genres: string;
  }[];
}

// Kind of a workaround for https://github.com/awslabs/dynamodb-data-mapper-js/issues/136
export type HistoryRequired = Required<
  Omit<History, 'type' | 'created_at' | 'expire_at'>
>;

// Store dates as ISO strings.

// Note that for the TTL attribute, this decorator cannot be applied
// as it needs to be stored as a Number data type.
// https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/time-to-live-ttl-before-you-start.html#time-to-live-ttl-before-you-start-formatting
//
const ISOdateType: CustomType<Date> = {
  type: 'Custom',
  marshall: (input: Date): AttributeValue => ({ S: input.toISOString() }),
  /* eslint-disable @typescript-eslint/no-non-null-assertion */
  unmarshall: (persistedValue: AttributeValue): Date =>
    new Date(persistedValue.S!),
  /* eslint-enable */
};

@table(config.AWS_TABLE_NAME)
export class History {
  @hashKey({ defaultProvider: () => 'history' })
  type?: string;

  /* Songs expire after 30 days by default */
  @attribute({
    defaultProvider: () =>
      moment().add(config.dynamoExpireAfter, 'days').toDate(),
  })
  expire_at?: Date;

  /* Date of the last song that was listened to */
  @rangeKey(ISOdateType)
  timestamp?: Date;

  /* Date of when the songs were scrobbed */
  @attribute({
    ...ISOdateType,
    defaultProvider: () => new Date(),
  })
  created_at?: Date;

  /* Number of songs */
  @attribute()
  count?: number;

  /* Actual songs */
  @attribute()
  songs?: HistoryElement[];
}
