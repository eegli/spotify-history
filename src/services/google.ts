import { google } from 'googleapis';
import config from '../config';

const oAuth2Client = new google.auth.OAuth2(
  config.GOOGLE.client_id,
  config.GOOGLE.client_secret
);

oAuth2Client.setCredentials({ refresh_token: config.GOOGLE.refresh_token });

export const drive = google.drive({ version: 'v3', auth: oAuth2Client });
