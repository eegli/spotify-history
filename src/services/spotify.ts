import { initialize, SpotifyClient } from '@spotifly/core';
import config from '../config';
import { DynamoHistoryElement } from '../config/types';

export type HistoryParams = Parameters<
  SpotifyClient['Player']['getRecentlyPlayedTracks']
>[0];

export interface RefreshTokenResponse {
  access_token: string;
  token_type: string;
  scope: string;
  expires_in: number;
}

// Export so we can use it in the tests
export type HistoryResponse = SpotifyApi.UsersRecentlyPlayedTracksResponse;
export type MultipleArtistsResponse = SpotifyApi.MultipleArtistsResponse;

export default class Spotify {
  private client: SpotifyClient;
  private items: SpotifyApi.PlayHistoryObject[] = [];
  // Unix timestamps
  cursorBefore = 0;
  cursorAfter = 0;

  constructor() {
    this.client = initialize({
      clientId: config.SPOTIFY.client_id,
      clientSecret: config.SPOTIFY.client_secret,
      refreshToken: config.SPOTIFY.refresh_token,
    });
  }

  get count() {
    return this.items.length;
  }

  // Gets the raw history from Spotify
  async fetchHistory(params: HistoryParams): Promise<void> {
    const res = await this.client.Player.getRecentlyPlayedTracks({
      ...params,
      limit: 50,
    });

    if (res.data.items.length > 0) {
      this.items = res.data.items;

      const firstElem = this.items[0].played_at;
      const lastElem = this.items[this.items.length - 1].played_at;

      // Manually set the cursors so that we know the time of the
      // first and last song of this history
      this.cursorAfter = new Date(lastElem).getTime();
      this.cursorBefore = new Date(firstElem).getTime();
    }
  }

  // Create the actual history that will be saved in dynamo
  createHistory(): DynamoHistoryElement[] {
    return this.items.reduce((acc, el) => {
      acc.push({
        name: el.track.name,
        id: el.track.id,
        playedAt: new Date(el.played_at).toISOString(),
      });

      return acc;
    }, <DynamoHistoryElement[]>[]);
  }
}
