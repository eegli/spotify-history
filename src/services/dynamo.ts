import config from '../config';

import { DataMapper, QueryOptions } from '@aws/dynamodb-data-mapper';
import { ConditionExpression } from '@aws/dynamodb-expressions';
import DynamoDB from 'aws-sdk/clients/dynamodb';

import { ServiceConfigurationOptions } from 'aws-sdk/lib/service';
import { History, HistoryElement } from '../models/history';

type Options = DynamoDB.DocumentClient.DocumentClientOptions &
  ServiceConfigurationOptions &
  DynamoDB.ClientApiVersions;

const params: Options = {
  region: config.AWS_REGION,
  apiVersion: '2012-08-10',
};

if (process.env.stage === 'local') {
  params.region = 'localhost';
  params.endpoint = 'http://localhost:8000';
}

const mapper = new DataMapper({
  client: new DynamoDB(params),
});

export const dynamoGetLatestHistory = async () => {
  const queryOptions: QueryOptions = {
    limit: 1,
    scanIndexForward: false,
  };

  const params: ConditionExpression = {
    type: 'Equals',
    subject: 'type',
    object: 'history',
  };
  const iterator = mapper.query(History, params, queryOptions);
  for await (const history of iterator) {
    // Return the first element
    return history;
  }
};

export const dynamoSetHistory = async (
  timestamp: string,
  count: number,
  songs: HistoryElement[]
) => {
  const newHistory: History = Object.assign(new History(), {
    timestamp,
    count,
    songs,
  });

  return mapper.put(newHistory);
};
