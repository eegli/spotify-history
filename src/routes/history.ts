import { QueryOptions } from '@aws/dynamodb-data-mapper';
import { AndExpression, ConditionExpression } from '@aws/dynamodb-expressions';
import moment from 'moment';
import { defaults } from '../config';
import { DynamoHistoryElement } from '../config/types';
import { History, HistoryRequired } from '../models/history';
import { dynamoDataMapper } from '../services/dynamo';

const params: ConditionExpression = {
  type: 'Equals',
  subject: 'type',
  object: 'history',
};
export const dynamoGetLatestHistory = async () => {
  const queryOptions: QueryOptions = {
    limit: 1,
    scanIndexForward: false,
  };

  const iterator = dynamoDataMapper.query(History, params, queryOptions);
  for await (const history of iterator) {
    // Return the first element
    return history;
  }
};

export const dynamoSetHistory = async ({
  timestamp,
  songs,
}: HistoryRequired) => {
  const newHistory: History = Object.assign(new History(), {
    timestamp,
    songs,
  });
  return dynamoDataMapper.put(newHistory);
};

export const dynamoGetHistoryRange = async () => {
  // For comparison, the timestamp needs to be an ISO string just like
  // the timestamp from the model
  const timestamp = moment()
    .subtract(...defaults.backupRange)
    .toISOString();

  const dateFilter: ConditionExpression = {
    type: 'GreaterThanOrEqualTo',
    subject: 'timestamp',
    object: timestamp,
  };

  const filters: AndExpression = {
    type: 'And',
    conditions: [params, dateFilter],
  };

  const items: History[] = [];

  const iterator = dynamoDataMapper.query(History, filters);

  for await (const history of iterator) {
    items.push(history);
  }

  // Reduce the history to only get the songs
  return items.reduce((acc, curr) => {
    if (curr.songs) {
      acc.push(...curr.songs);
    }
    return acc;
  }, <DynamoHistoryElement[]>[]);
};
