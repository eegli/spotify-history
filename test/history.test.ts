import { QueryIterator, StringToAnyObjectMap } from '@aws/dynamodb-data-mapper';
import { Context, EventBridgeEvent } from 'aws-lambda';
import axios from 'axios';
import { handler } from '../src/handler.history';
import { dynamoDataMapper } from '../src/services/dynamo';
import Spotify, { HistoryParams } from '../src/services/spotify';
import {
  dynamoFullData,
  emptyResponse,
  spotifyHistoryResponse,
  spotifyPutTokenResponse,
} from './payloads';

// Mock the dynamo client so that we don't accidentally hit the api
jest.mock('../src/services/dynamo');
jest.mock('@aws/dynamodb-data-mapper');
jest.mock('axios');

const querySpy = jest.spyOn(dynamoDataMapper, 'query');
const putSpy = jest.spyOn(dynamoDataMapper, 'put');

// This will be overwritten sometimes
const iteratorMock = jest.fn().mockReturnValue([].values());

// @ts-ignore
const qi: QueryIterator<StringToAnyObjectMap> = QueryIterator;
// @ts-ignore
qi[Symbol.iterator] = iteratorMock;

querySpy.mockReturnValue(qi);

const mockAxios = axios as jest.Mocked<typeof axios>;
const fetchSpotifySpy = jest.spyOn(Spotify.prototype, 'fetchHistory');

beforeEach(() => {
  mockAxios.post.mockResolvedValue(spotifyPutTokenResponse);

  mockAxios.get.mockResolvedValueOnce(spotifyHistoryResponse);

  fetchSpotifySpy.mockClear();
  querySpy.mockClear();
  putSpy.mockClear();
  iteratorMock.mockClear();
});

describe('Handler', () => {
  it('runs without errors', async () => {
    await handler(
      {} as EventBridgeEvent<'Scheduled Event', any>,
      {} as Context,
      () => {}
    );
    expect(querySpy).toHaveBeenCalledTimes(1);
    expect(putSpy).toHaveBeenCalledTimes(1);
    expect(putSpy.mock.calls).toMatchSnapshot();
  });

  it('queries all songs if there are no existing ones', async () => {
    await handler(
      {} as EventBridgeEvent<'Scheduled Event', any>,
      {} as Context,
      () => {}
    );

    expect(fetchSpotifySpy.mock.calls[0][0]).toMatchObject<HistoryParams>({
      before: expect.any(Number),
      limit: expect.any(Number),
    });
  });

  it('only queries new songs if there are existing ones', async () => {
    iteratorMock.mockReturnValueOnce(dynamoFullData().values());
    await handler(
      {} as EventBridgeEvent<'Scheduled Event', any>,
      {} as Context,
      () => {}
    );
    expect(fetchSpotifySpy.mock.calls[0][0]).toMatchObject<HistoryParams>({
      after: expect.any(Number),
      limit: expect.any(Number),
    });
  });

  it('saves nothing to dynamo if there is nothing new', async () => {
    // Overwrite initial history request, mock "no new songs"
    mockAxios.get = jest.fn().mockResolvedValueOnce(emptyResponse);

    await handler(
      {} as EventBridgeEvent<'Scheduled Event', any>,
      {} as Context,
      () => {}
    );
    expect(putSpy).not.toHaveBeenCalled();
  });
});
