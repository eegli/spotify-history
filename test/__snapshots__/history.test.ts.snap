// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Handler runs without errors 1`] = `
Array [
  Array [
    History {
      "count": 3,
      "songs": Array [
        Object {
          "artists": Array [
            Object {
              "id": "3rVvL036o05Y34SPRsW63P",
              "name": "Elfenberg",
            },
          ],
          "id": "2HuWC2OLujMmdm1Mut8rjM",
          "name": "Loma",
          "playedAt": "2021-09-08T07:23:11.995Z",
        },
        Object {
          "artists": Array [
            Object {
              "id": "0rasA5Z5h1ITtHelCpfu9R",
              "name": "Paul Kalkbrenner",
            },
          ],
          "id": "33R7eq2Yqrhjz7eP6XhCsy",
          "name": "Graf Zahl",
          "playedAt": "2021-09-07T19:12:49.127Z",
        },
        Object {
          "artists": Array [
            Object {
              "id": "54KTsBl98buzGkELjGlnuU",
              "name": "Active Child",
            },
            Object {
              "id": "1ehDqG6bFGAurxlAJeQgNT",
              "name": "Matthew Dekay",
            },
          ],
          "id": "3SIIAdWl4RmLNYpppSUy7b",
          "name": "Johnny Belinda - Matthew Dekay Remix",
          "playedAt": "2021-09-08T07:27:05.283Z",
        },
      ],
      "timestamp": 2021-09-08T07:23:11.995Z,
    },
  ],
]
`;
